File: backend/workouts.py
from flask import Blueprint, request, jsonify, session
from models import Workout
from bson import ObjectId

workouts_bp = Blueprint('workouts', __name__, url_prefix='/api/workouts')

def require_auth(f):
    """Decorator to require authentication"""
    def decorated_function(*args, **kwargs):
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({'error': 'Authentication required'}), 401
        return f(user_id, *args, **kwargs)
    decorated_function.__name__ = f.__name__
    return decorated_function

def serialize_workout(workout):
    """Convert workout object to JSON-serializable format"""
    if workout:
        return {
            'id': str(workout['_id']),
            'user_id': str(workout['user_id']),
            'discipline': workout['discipline'],
            'duration': workout['duration'],
            'intensity': workout['intensity'],
            'notes': workout['notes'],
            'date': workout['date'],
            'created_at': workout['created_at'].isoformat(),
            'updated_at': workout['updated_at'].isoformat()
        }
    return None

@workouts_bp.route('', methods=['GET'])
@require_auth
def get_workouts(user_id):
    """Get all workouts for the authenticated user"""
    try:
        limit = int(request.args.get('limit', 100))
        skip = int(request.args.get('skip', 0))

        workouts = Workout.find_by_user(user_id, limit=limit, skip=skip)
        return jsonify({
            'workouts': [serialize_workout(w) for w in workouts]
        }), 200

    except Exception as e:
        print(f"Get workouts error: {e}")
        return jsonify({'error': 'Failed to fetch workouts'}), 500


@workouts_bp.route('', methods=['POST'])
@require_auth
def create_workout(user_id):
    """Create a new workout"""
    try:
        data = request.get_json()

        # Validate input
        required_fields = ['discipline', 'duration', 'intensity']
        if not data or not all(k in data for k in required_fields):
            return jsonify({'error': 'Missing required fields'}), 400

        discipline = data['discipline']
        duration = int(data['duration'])
        intensity = int(data['intensity'])
        notes = data.get('notes', '')
        date = data.get('date')

        # Validate values
        if duration <= 0 or duration > 1440:  # Max 24 hours
            return jsonify({'error': 'Duration must be between 1 and 1440 minutes'}), 400

        if intensity < 1 or intensity > 10:
            return jsonify({'error': 'Intensity must be between 1 and 10'}), 400

        # Create workout
        workout = Workout.create(user_id, discipline, duration, intensity, notes, date)

        return jsonify({
            'message': 'Workout created successfully',
            'workout': serialize_workout(workout)
        }), 201

    except ValueError as e:
        return jsonify({'error': 'Invalid input values'}), 400
    except Exception as e:
        print(f"Create workout error: {e}")
        return jsonify({'error': 'Failed to create workout'}), 500


@workouts_bp.route('/<workout_id>', methods=['GET'])
@require_auth
def get_workout(user_id, workout_id):
    """Get a specific workout"""
    try:
        workout = Workout.find_by_id(workout_id, user_id)

        if not workout:
            return jsonify({'error': 'Workout not found'}), 404

        return jsonify({'workout': serialize_workout(workout)}), 200

    except Exception as e:
        print(f"Get workout error: {e}")
        return jsonify({'error': 'Failed to fetch workout'}), 500


@workouts_bp.route('/<workout_id>', methods=['PUT'])
@require_auth
def update_workout(user_id, workout_id):
    """Update a workout"""
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No data provided'}), 400

        # Build update data
        update_data = {}
        if 'discipline' in data:
            update_data['discipline'] = data['discipline']
        if 'duration' in data:
            duration = int(data['duration'])
            if duration <= 0 or duration > 1440:
                return jsonify({'error': 'Duration must be between 1 and 1440 minutes'}), 400
            update_data['duration'] = duration
        if 'intensity' in data:
            intensity = int(data['intensity'])
            if intensity < 1 or intensity > 10:
                return jsonify({'error': 'Intensity must be between 1 and 10'}), 400
            update_data['intensity'] = intensity
        if 'notes' in data:
            update_data['notes'] = data['notes']
        if 'date' in data:
            update_data['date'] = data['date']

        if not update_data:
            return jsonify({'error': 'No valid fields to update'}), 400

        # Update workout
        success = Workout.update(workout_id, user_id, update_data)

        if not success:
            return jsonify({'error': 'Workout not found or update failed'}), 404

        # Fetch updated workout
        workout = Workout.find_by_id(workout_id, user_id)

        return jsonify({
            'message': 'Workout updated successfully',
            'workout': serialize_workout(workout)
        }), 200

    except ValueError as e:
        return jsonify({'error': 'Invalid input values'}), 400
    except Exception as e:
        print(f"Update workout error: {e}")
        return jsonify({'error': 'Failed to update workout'}), 500


@workouts_bp.route('/<workout_id>', methods=['DELETE'])
@require_auth
def delete_workout(user_id, workout_id):
    """Delete a workout"""
    try:
        success = Workout.delete(workout_id, user_id)

        if not success:
            return jsonify({'error': 'Workout not found'}), 404

        return jsonify({'message': 'Workout deleted successfully'}), 200

    except Exception as e:
        print(f"Delete workout error: {e}")
        return jsonify({'error': 'Failed to delete workout'}), 500


@workouts_bp.route('/stats', methods=['GET'])
@require_auth
def get_stats(user_id):
    """Get workout statistics for the authenticated user"""
    try:
        stats = Workout.get_stats(user_id)

        if not stats:
            return jsonify({
                'total_sessions': 0,
                'total_hours': 0,
                'avg_intensity': 0
            }), 200

        return jsonify({
            'total_sessions': stats['total_sessions'],
            'total_hours': round(stats['total_duration'] / 60, 1),
            'avg_intensity': round(stats['avg_intensity'], 1)
        }), 200

    except Exception as e:
        print(f"Get stats error: {e}")
        return jsonify({'error': 'Failed to fetch statistics'}), 500

File: backend-ts/src/routes/workouts.ts
import { Router, Request, Response } from 'express';
import { Workout } from '../models/Workout';
import { requireAuth } from '../middleware/auth';
import { Server as SocketIOServer } from 'socket.io';

const router = Router();

// Socket.io instance will be injected
let io: SocketIOServer;

export const setSocketIO = (socketIO: SocketIOServer) => {
  io = socketIO;
};

// Get all workouts for authenticated user
router.get('/', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;
    const limit = parseInt(req.query.limit as string) || 100;
    const skip = parseInt(req.query.skip as string) || 0;

    const workouts = await Workout
      .find({ userId })
      .sort({ date: -1, createdAt: -1 })
      .skip(skip)
      .limit(limit)
      .lean();

    const formattedWorkouts = workouts.map(w => ({
      id: w._id.toString(),
      discipline: w.discipline,
      duration: w.duration,
      intensity: w.intensity,
      notes: w.notes,
      date: w.date,
      createdAt: w.createdAt,
      updatedAt: w.updatedAt
    }));

    res.status(200).json({
      workouts: formattedWorkouts
    });
  } catch (error) {
    console.error('Get workouts error:', error);
    res.status(500).json({ error: 'Failed to get workouts' });
  }
});

// Create new workout
router.post('/', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;
    const { discipline, duration, intensity, notes, date } = req.body;

    // Validate input
    if (!discipline || !duration || !intensity) {
      res.status(400).json({ error: 'Missing required fields' });
      return;
    }

    if (duration < 1) {
      res.status(400).json({ error: 'Duration must be at least 1 minute' });
      return;
    }

    if (intensity < 1 || intensity > 10) {
      res.status(400).json({ error: 'Intensity must be between 1 and 10' });
      return;
    }

    // Create workout
    const workout = await Workout.create({
      userId,
      discipline,
      duration: parseInt(duration),
      intensity: parseInt(intensity),
      notes: notes || '',
      date: date || new Date().toISOString().split('T')[0]
    });

    const formattedWorkout = {
      id: workout._id.toString(),
      discipline: workout.discipline,
      duration: workout.duration,
      intensity: workout.intensity,
      notes: workout.notes,
      date: workout.date,
      createdAt: workout.createdAt,
      updatedAt: workout.updatedAt
    };

    // Emit real-time update to all connected clients for this user
    if (io) {
      io.to(`user:${userId}`).emit('workout:created', formattedWorkout);
    }

    res.status(201).json({
      message: 'Workout created successfully',
      workout: formattedWorkout
    });
  } catch (error) {
    console.error('Create workout error:', error);
    res.status(500).json({ error: 'Failed to create workout' });
  }
});

// Get specific workout
router.get('/:id', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;
    const workoutId = req.params.id;

    const workout = await Workout.findOne({
      _id: workoutId,
      userId
    });

    if (!workout) {
      res.status(404).json({ error: 'Workout not found' });
      return;
    }

    res.status(200).json({
      workout: {
        id: workout._id.toString(),
        discipline: workout.discipline,
        duration: workout.duration,
        intensity: workout.intensity,
        notes: workout.notes,
        date: workout.date,
        createdAt: workout.createdAt,
        updatedAt: workout.updatedAt
      }
    });
  } catch (error) {
    console.error('Get workout error:', error);
    res.status(500).json({ error: 'Failed to get workout' });
  }
});

// Update workout
router.put('/:id', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;
    const workoutId = req.params.id;
    const { discipline, duration, intensity, notes, date } = req.body;

    const updateData: any = {};
    if (discipline !== undefined) updateData.discipline = discipline;
    if (duration !== undefined) updateData.duration = parseInt(duration);
    if (intensity !== undefined) updateData.intensity = parseInt(intensity);
    if (notes !== undefined) updateData.notes = notes;
    if (date !== undefined) updateData.date = date;

    const workout = await Workout.findOneAndUpdate(
      { _id: workoutId, userId },
      updateData,
      { new: true }
    );

    if (!workout) {
      res.status(404).json({ error: 'Workout not found' });
      return;
    }

    const formattedWorkout = {
      id: workout._id.toString(),
      discipline: workout.discipline,
      duration: workout.duration,
      intensity: workout.intensity,
      notes: workout.notes,
      date: workout.date,
      createdAt: workout.createdAt,
      updatedAt: workout.updatedAt
    };

    // Emit real-time update
    if (io) {
      io.to(`user:${userId}`).emit('workout:updated', formattedWorkout);
    }

    res.status(200).json({
      message: 'Workout updated successfully',
      workout: formattedWorkout
    });
  } catch (error) {
    console.error('Update workout error:', error);
    res.status(500).json({ error: 'Failed to update workout' });
  }
});

// Delete workout
router.delete('/:id', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;
    const workoutId = req.params.id;

    const workout = await Workout.findOneAndDelete({
      _id: workoutId,
      userId
    });

    if (!workout) {
      res.status(404).json({ error: 'Workout not found' });
      return;
    }

    // Emit real-time update
    if (io) {
      io.to(`user:${userId}`).emit('workout:deleted', { id: workoutId });
    }

    res.status(200).json({
      message: 'Workout deleted successfully'
    });
  } catch (error) {
    console.error('Delete workout error:', error);
    res.status(500).json({ error: 'Failed to delete workout' });
  }
});

// Get workout statistics
router.get('/stats/summary', requireAuth, async (req: Request, res: Response): Promise<void> => {
  try {
    const userId = req.session.userId!;

    const stats = await (Workout as any).getStats(userId);

    res.status(200).json({
      stats: stats || {
        totalSessions: 0,
        totalDuration: 0,
        avgIntensity: 0
      }
    });
  } catch (error) {
    console.error('Get stats error:', error);
    res.status(500).json({ error: 'Failed to get statistics' });
  }
});

export default router;
